File: dailyQuestion/2020-05/05-27/python/solution_floyd.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

def detectCycle(head: ListNode) -> ListNode:
    def getInstance(head):
        fast = head
        slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:
                return slow
        return None

    if not head:
        return None
    prt2 = getInstance(head)
    if not prt2:
        return None

    prt1 = head
    while prt1 != prt2:
        prt1 = prt1.next
        prt2 = prt2.next
    return prt1


if __name__ == "__main__":
    node = ListNode(3)
    node.next = ListNode(2)
    node.next.next = ListNode(0)
    node.next.next.next = ListNode(-4)
    node.next.next.next.next = node.next
    result = detectCycle(node)
    print(result.val)

File: dailyQuestion/2020/2020-05/05-27/golang/solution_floyd.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {
	node := &ListNode{Val: 3, Next: &ListNode{Val: 2}}
	node.Next.Next = &ListNode{Val: 0}
	node.Next.Next.Next = &ListNode{Val: -4}
	node.Next.Next.Next.Next = node.Next
	result := detectCycle(node)
	fmt.Println(result)

}

func detectCycle(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}
	prt2 := getInstances(head)
	if prt2 == nil {
		return nil
	}
	prt1 := head
	if prt1 != prt2 {
		prt1 = prt1.Next
		prt2 = prt2.Next
	}
	return prt1
}

func getInstances(head *ListNode) *ListNode {
	slow := head
	fast := head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast {
			return slow
		}
	}
	return nil
}
